File: HR-python/shortestReach.py
import queue
class Graph:
    def __init__(self, n):
        self.graph = dict()
        self.n = n
        for i in range(n):
            self.graph[i] = []

    def connect(self, node_1, node_2): #undirected graph
        self.graph[node_1].append(node_2)
        self.graph[node_2].append(node_1)

    def find_all_distances(self, source_node):
        q = queue.Queue()
        q.put(source_node)
        visited = dict()
        level = 0
        no_of_nodes_at_this_level = 1
        distances = dict()
        while not q.empty():
            node = q.get()
            no_of_nodes_at_this_level -= 1

            if not node in visited:
                for i in range(len(self.graph[node])):
                    if not self.graph[node][i] in visited:
                        q.put(self.graph[node][i])
                distances[node] = level
                visited[node] = True

            if no_of_nodes_at_this_level is 0:
                no_of_nodes_at_this_level = q.qsize()
                level += 1

        return distances
    
    def print_distances(self, distances, source_node):
        distance = []
        for i in range(self.n):
            if (i is not source_node and i in distances):
                distance.append(distances[i] * 6)
            elif i is not source_node:
                distance.append(-1)
        for i in range(len(distance)):
            print(distance[i], end=" ")
        print("")
                        

def main():
    graph = Graph(4)
    graph.connect(0, 1)
    graph.connect(0, 2)
    s = 1
    distances = graph.find_all_distances(s-1)
    graph.print_distances(distances, s-1)


main()

File: cpp/shortestReach.cpp
#include<bits/stdc++.h>
using namespace std;

class Graph {
    public: 
        vector<vector<int>> graph;
        int n = 0;

        Graph(int n) {
            this -> n = n;
            graph.resize(n);
        }

        void add_edge(int u, int v) {
            graph[u].push_back(v);
            graph[v].push_back(u);
        }

        vector<int> shortest_reach(int start) {
            queue<int> q;
            q.push(start);
            int level = 0;
            int noOfNodesAtThisLevel = 1;
            vector<bool> visited(n, false);
            map<int, int> distance;

            while(!q.empty()) {
                int node = q.front();
                q.pop();
                noOfNodesAtThisLevel--;
                
                if (!visited[node]) {
                    for (int i = 0; i < graph[node].size(); i++) {
                        q.push(graph[node][i]);
                    }
                    visited[node] = true;
                    distance.insert(pair <int, int> (node, level));
                }
                if (!noOfNodesAtThisLevel) { // == 0 -> false
                    noOfNodesAtThisLevel = q.size();
                    level++;
                }
            }
            vector<int> distanceVector = get_distance_vector(distance, start);
            return distanceVector;
        }
        vector<int> get_distance_vector(map<int, int> &distance, int start) {
            vector<int> distanceVector;
            for (int i = 0; i < n; i++) {
                map<int, int>::iterator it = distance.find(i);
                if (i != start && it != distance.end()) {
                    int d = (distance.find(i) -> second) * 6;
                    distanceVector.push_back(d);
                } else if (i != start) {
                    distanceVector.push_back(-1);
                }
            }
            return distanceVector;
        }
};

int main() {
    int queries = 1;
        
    for (int t = 0; t < queries; t++) {
		int n = 4, m = 2;
        Graph graph(n);
        graph.add_edge(0, 1);
        graph.add_edge(0, 2);
		int startId = 0;
        vector<int> distances = graph.shortest_reach(startId);

        for (int i = 0; i < distances.size(); i++) {
            cout << distances[i] << " ";
        }
        cout << endl;
    }
    
    return 0;
}